### ISSUE TITLE
VMR repositories should receive configuration properties like OS, Arch and RID and respect them

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/source-build#4784]
The .NET SDK targets a specific RID, i.e. `win-x64`. When that SDK is being used in a vertical that should produce `win-x86` assets, the wrong assets are restored.

Any component that restores RID specific assets needs to know the RID that the VMR is building for. That means any project that sets `<OutputType>Exe</OutputType>` or similarly WinExe, SelfContained, publishes, etc. RID specific assets are apphost packs, runtime packs and many more.

We need to the following things:
1. Pass a set of properties into every VMR repo's build. Initially `TargetOS`, `TargetArchitecture` and `TargetRuntimeIdentifier` might be enough.
2. Repositories should respect those properties. Probably the best way to achieve this is by leveraging the Arcade SDK to read from those properties and set the correct defaults. See the example below.

Arcade SDK:
```xml
<PropertyGroup>
  <!-- source-build sets the following already in a few repos that depend on live runtime, i.e. aspnetcore. -->
  <DefaultAppHostRuntimeIdentifier>$(TargetRuntimeIdentifier)</DefaultAppHostRuntimeIdentifier>

  <!-- the following might be better as it affects all packs, not just the apphost ones: -->
  <NETCoreSdkPortableRuntimeIdentifier>$(TargetRuntimeIdentifier)</NETCoreSdkPortableRuntimeIdentifier>
  <NETCoreSdkRuntimeIdentifier>$(TargetRuntimeIdentifier)</NETCoreSdkRuntimeIdentifier >
</PropertyGroup>
```

There are probably more properties that should be set that control what RID to use for restoring packs. @dsplaisted do we have a list of those?

**Comment by ViktorHofer** [comment]
I wonder if overriding `NETCoreSdkPortableRuntimeIdentifier` and `NETCoreSdkRuntimeIdentifier` is supported so that the SDK only downloadings assets with that RID.

**Comment by mmitche** [comment]
I don't think this is the case for all repos. I think this is only for repos that depend on the live runtime AND produce rid specific assets. Roslyn is not in this category, for instance, and should not be. The assets it produces (including an exes) are not rid specific.

**Comment by ViktorHofer** [comment]
> The assets it produces (including an exes) are not rid specific.

The apphost which is included for `OutputType=Exe` by default (unless you set `UseAppHost=false`) is a RID specific asset. Also see the SDK definition of when a project is RID agnostic: https://github.com/dotnet/sdk/blob/363ff49b2cbed4e6978c40750c5e59b3a1fc56c6/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L117C35-L117C48

This is really easy to reproduce locally. Just create a `dotnet new console` project and look at the app output. The binlog will also show you that a PackageDownload item exists for the apphost packages (clear you `C:\Program Files\dotnet\packs` folder before you do this though, otherwise NuGet doesn't get called).

---

OK, now bear with me with a pretty verbose description of what's going on:

Note that this isn't a VMR specific problem. There's a mismatch whenever the host RID is different from the target RID. In [GNU configure terms](https://gcc.gnu.org/onlinedocs/gccint/Configure-Terms.html) this is called a `cross` configuration.
Example: I'm on an **win-x64** OS (called build) and use **win-x64** SDK (called host). I use that to restore my project (called target) which sets OutputType=Exe or SelfContained=true (or any other property that results in RID specific assets getting downloaded). I want my project to only restore `win-x86` RID assets and I don't specify `RuntimeIdentifier(s)`, I let the SDK handle this for me.

In that configuration you download `win-x64` RID assets because that's the host's RID. If you want your target to only ever downlodad `win-x86` RID specific assets (if necessary), you apparently need to override `NETCoreSdkPortableRuntimeIdentifier` and `NETCoreSdkRuntimeIdentifier` properties.

For the VMR we definitely care about this problem as a win-x86 vertical should depend on `win-x86` assets. Arcade or any other repository that doesn't depend on live runtime isn't failing because those repos automatically use the RID packs that are available in the SDK's packs folder (pre-cached). For repositories that depend on live runtime, @pavel-purma's build which uses a unique OfficialBuildId demonstrates that we are now trying to restore RID packs for the wrong RID.

As an example, take the Microsoft.DotNet.Tar project which sets `OutputType=Exe` but doesn't set a RuntimeIdentifier. When building that in the win-x86 vertical, still the win-x64 apphost gets picked:

![Image](https://github.com/user-attachments/assets/8eea2d11-db6d-4820-8fd6-ad0611c8a787)

This is often not a problem _at consumption time_ as we usually don't use an app's apphost directly and invoke the executable with the `dotnet` host instead. Also, `PackAsTool` or NuGet Pack by default don't include the apphost.

cc @baronfel @richlander to fact check me here

**Comment by mmitche** [comment]
Got it, that makes sense.

If you set `RuntimeIdentifier` to win-x86 when building using the x64 SDK, do you end up downloading x64 apphost as well as the x86 apphost? Or does it properly constrain the SDK?

**Comment by ViktorHofer** [comment]
The `RuntimeIdentifier` / `RuntimeIdentifiers` property wins over any SDK default. That's why we had to make these changes in couple of repos: https://github.com/dotnet/aspnetcore/blob/849235790fe151ab311192d7e6c5860e6c177f5e/src/Servers/testassets/ServerComparison.TestSites/ServerComparison.TestSites.csproj#L8

**Comment by ViktorHofer** [comment]
This can be closed now. The Target properties are passed to the inner repos already and we have the RID filtering infra in Arcade now.
